### PR TITLE
Add kci_rootfs command line tool

### PR DESCRIPTION
**Pull request by mgalka**
Add basic initial implementation of the `kci_rootfs` tool which includes:
- configuration parser
- simple buster debos based rootfs configuration in rootfs-configs.yaml
- `kci_rootfs validate` subcommand that lists all configuration data based on the YAML file
- `kci_rootfs list_configs` subcommand that lists all rootfs configs available in the YAML file

### CODE REVIEW

File: kci_rootfs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
```
Comment (gctucker) [inline kci_rootfs]: This should be Python 3.

File: kci_rootfs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        rootfs_configs = config_data['rootfs_configs']
+        for config_name in rootfs_configs.keys():
+            print(config_name)
+
+# -----------------------------------------------------------------------------
+# Main
+#
+
+
+if __name__ == '__main__':
```
Comment (gctucker) [inline kci_rootfs]: You can use the default main function:
```
if __name__ == '__main__':
    args = parse_args("kci_build", "rootfs-configs.yaml", globals())
    configs = kernelci.config.rootfs.from_yaml(args.yaml_configs)
    status = args.func(configs, args)
    sys.exit(0 if status is True else 1)
```

File: kci_rootfs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        for config_name, config in rootfs_configs.items():
+            print('\trootfs_type: {}'.format(config.rootfs_type))
+            print('\tarch_list: {}'.format(config.arch_list))
+            print('\tdebian_release: {}'.format(config.debian_release))
+            print('\textra_packages: {}'.format(config.extra_packages))
+            print('\textra_packages_remove: {}'
+                  .format(config.extra_packages_remove))
+            print('\textra_files_remove: {}'.format(config.extra_files_remove))
+            print('\tscript: {}'.format(config.script))
+            print('\tdocker_image: {}'.format(config.docker_image))
```
Comment (gctucker) [inline kci_rootfs]: Add:
```python
return True
```
otherwise the status code will be non-zero.

File: kci_rootfs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from kernelci.cli import Args, Command
+import kernelci.config.rootfs
+
+# -----------------------------------------------------------------------------
+# Commands
+#
+
+
+class cmd_validate(Command):
+    help = "Lists all rootfs configs"
```
Comment (gctucker) [inline kci_rootfs]: This should be rather something like "Validate the rootfs YAML configuration"

File: kci_rootfs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            print('\tdocker_image: {}'.format(config.docker_image))
+
+
+class cmd_list_configs(Command):
+    help = "Lists all rootfs config names"
+
+    def __call__(self, config_data, *args, **kwargs):
+        rootfs_configs = config_data['rootfs_configs']
+        for config_name in rootfs_configs.keys():
+            print(config_name)
```
Comment (gctucker) [inline kci_rootfs]: Same issue here:
```python
return True
```

File: kci_rootfs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# Commands
+#
+
+
+class cmd_validate(Command):
+    help = "Lists all rootfs configs"
+
+    def __call__(self, config_data, *args, **kwargs):
+        rootfs_configs = config_data['rootfs_configs']
+        print(rootfs_configs)
```
Comment (gctucker) [inline kci_rootfs]: I think the `validate` command should be more about validating the YAML data than showing things, there can be other commands for that such as `list_configs`.  What's important is to return an error if the YAML doesn't follow the schema, so we can also run it in Travis CI.

File: kernelci/config/rootfs.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                 docker_image=None):
+
+        super(RootFS_Debos, self).__init__(name, rootfs_type)
+        self._debian_release = debian_release
+        self._arch_list = arch_list or list()
+        self._extra_packages = extra_packages or list()
+        self._extra_packages_remove = extra_packages_remove or list()
+        self._extra_files_remove = extra_files_remove or list()
+        self._script = script
+        self._docker_image = docker_image
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: Please remove references to Docker from this config, the rootfs doesn't have anything to do with Docker.  Docker can be used to produce the rootfs but it's not a property of it.

File: kernelci/config/rootfs.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    def debian_release(self):
+        return self._debian_release
+
+    @property
+    def arch_list(self):
+        return list(self._arch_list)
+
+    @property
+    def extra_packages(self):
+        return ' '.join(list(self._extra_packages))
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: Is there any value in converting this list into a string?  Maybe returning a list would be better and the caller can convert it to a string if needed.  The config in YAML is a list, this is a list of packages so I think a list makes more sense here.

File: kernelci/config/rootfs.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    @property
+    def extra_packages_remove(self):
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: Same thing as `extra_packages`.

File: kernelci/config/rootfs.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        return ' '.join(list(self._extra_packages_remove))
+
+    @property
+    def extra_files_remove(self):
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: Same thing as `extra_packages`.

File: kernelci/config/rootfs.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        return rootfs_cls.from_yaml(rootfs, kw)
+
+
+def from_yaml(yaml_path):
+    with open(yaml_path) as f:
+        data = yaml.safe_load(f)
+
+    rootfs_configs = {
+        name: RootFSFactory.from_yaml(name, rootfs)
+        for name, rootfs in data['rootfs_configs'].iteritems()
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: `.iteritems()` is a Python 2 thing, we're using Python 3 now.

File: rootfs-configs.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    extra_files_remove:
+      - '*networkd*'
+      - '*resolved*'
+      - tar
+      - patch
+      - dir
+      - partx
+      - find
+    script: ""
+    docker_image: godebos/debos
```
Comment (gctucker) [inline rootfs-configs.yaml]: As mentioned before, please remove references to Docker from this config.  Also, we're not currently using this image in KernelCI but `kernelci/base-debos`.

Comment (mgalka) [reply rootfs-configs.yaml]: removed

File: kernelci/config/rootfs.py (hunk 11 above)
Comment (mgalka) [reply kernelci/config/rootfs.py]: Obiously, for some reason I didn't take the Python3 migration into consideration. changed to `.items()`

File: kernelci/config/rootfs.py (hunk 10 above)
Comment (mgalka) [reply kernelci/config/rootfs.py]: fixed

File: kernelci/config/rootfs.py (hunk 9 above)
Comment (mgalka) [reply kernelci/config/rootfs.py]: fixed

File: kernelci/config/rootfs.py (hunk 8 above)
Comment (mgalka) [reply kernelci/config/rootfs.py]: Absolutely, I missed this one while preparing the PR. We should stick to returning non-processed data as much as possible.

File: kci_rootfs (hunk 5 above)
Comment (mgalka) [reply kci_rootfs]: fixed

File: kci_rootfs (hunk 4 above)
Comment (mgalka) [reply kci_rootfs]: fixed

File: kci_rootfs (hunk 3 above)
Comment (mgalka) [reply kci_rootfs]: fixed

File: kci_rootfs (hunk 2 above)
Comment (mgalka) [reply kci_rootfs]: Changed

File: kci_rootfs (hunk 1 above)
Comment (mgalka) [reply kci_rootfs]: Changed

File: kernelci/config/rootfs.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    def debian_release(self):
+        return self._debian_release
+
+    @property
+    def arch_list(self):
+        return list(self._arch_list)
+
+    @property
+    def extra_packages(self):
+        return self._extra_packages
```
Comment (gctucker) [inline kernelci/config/rootfs.py]: OK - except that by returning this list, it makes it possible to change it outside this class.  So maybe do like for `arch_list`:
```python
return list(self._extra_packages)
```

File: kernelci/config/rootfs.py (hunk 9 above)
Comment (gctucker) [reply kernelci/config/rootfs.py]: Same as above, a copy of the list avoids having it changed by code outside this class.

File: kernelci/config/rootfs.py (hunk 10 above)
Comment (gctucker) [reply kernelci/config/rootfs.py]: ditto

**Comment by gctucker** [comment]
@mgalka OK the code looks fine now, thanks.

However, this commit was written by Laks but it's under your name now:
```
commit 9e5c8b97fc4c9646dc4bf9ce4840ef769e79d73a
Author: Michal Galka <michal.galka@collabora.com>
Date:   Fri Nov 29 15:38:16 2019 +0100

    kci_rootfs: Add configuration for buster debos based rootfs
    
    Add configuration for a buster rootfs based on Jenkins jobs.
    Extend configuration parser and "validate" subcommand to work with
    new configuration.
    
    Signed-off-by: Lakshmipathi.G <lakshmipathi.ganapathi@collabora.com>
```

So I think the author should be reset to Laks, and he should be added to the list of authors in the copyright header.

Also some fixes were put in this commit but they should have been applied to the first commit in this PR `Add skeleton for kci_rootfs` instead.

**Comment by gctucker** [comment]
Thanks, LGTM and tested locally.

**Pull request merged**